Remove unnecessary ObjC object instantiation in JavaScriptEvaluationResult conversion to JSValueRef

achristensen07 opened the pull request:
#### 90da1bac176f6018de9d1ded7482bda821aa8e45
<pre>
Remove unnecessary ObjC object instantiation in JavaScriptEvaluationResult conversion to JSValueRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=291116">https://bugs.webkit.org/show_bug.cgi?id=291116</a>
<a href="https://rdar.apple.com/148630517">rdar://148630517</a>

Reviewed by Ryosuke Niwa.

This not only removes unnecessary work in the conversion process, but it also makes the code
able to be used on platforms that don&apos;t have ObjC.

Covered by many tests, especially AsyncFunction API tests and EvaluateJavaScript.ReturnTypes.

* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WTF::DefaultHash&lt;JSC::Strong&lt;JSC::JSCell&gt;&gt;::hash):
(WTF::DefaultHash&lt;JSC::Strong&lt;JSC::JSCell&gt;&gt;::equal):
(WebKit::JavaScriptEvaluationResult::toVariant):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::jsValueToVariant): Deleted.

Canonical link: <a href="https://commits.webkit.org/297138@main">https://commits.webkit.org/297138@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69b29ce7fac4f108c497fb277d398e4f885df911

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110667 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116693 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31005 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84150 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113615 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24762 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64591 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24125 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60488 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94151 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119483 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37707 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/28011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93114 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92938 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37963 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15706 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17856 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37603 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43075 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37265 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40604 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38973 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
```diff
@@ -288,55 +316,115 @@ static bool isSerializable(id argument)
+        return JSValueMakeNumber(context, value);
+    }, [&] (String&& value) -> JSValueRef {
+        auto string = OpaqueJSString::tryCreate(WTFMove(value));
+        return JSValueMakeString(context, string.get());
```
Why not just use `JSStringCreateWithUTF8CString` instead?

achristensen07 replied:
This avoids an unnecessary Latin1/UTF-16->UTF-8->Latin1/UTF-16 conversion roundtrip.

rniwa commented on Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
```diff
@@ -288,55 +316,115 @@ static bool isSerializable(id argument)
+        auto string = OpaqueJSString::tryCreate(WTFMove(value));
+        return JSValueMakeString(context, string.get());
+    }, [&] (Seconds value) -> JSValueRef {
+        JSValueRef argument = JSValueMakeNumber(context, value.value() * 1000.0);
```
Use `value.milliseconds()`?

rniwa commented on Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
```diff
@@ -288,55 +316,115 @@ static bool isSerializable(id argument)
+    for (auto [map, dictionary] : std::exchange(m_jsDictionaries, { })) {
+        for (auto [keyIdentifier, valueIdentifier] : map) {
+            ASSERT(JSValueIsString(context, m_instantiatedJSObjects.get(keyIdentifier)));
+            SUPPRESS_UNCOUNTED_ARG auto key = adopt(JSValueToStringCopy(context, m_instantiatedJSObjects.get(keyIdentifier), nullptr));
```
We should store `m_instantiatedJSObjects.get(keyIdentifier)` in a local Ref instead of suppressing the warning.

achristensen07 replied:
It is a JSValueRef which doesn't work with JSRetainPtr and even if it did we would need to keep SUPPRESS_UNCOUNTED_ARG.

achristensen07 committed:
Remove unnecessary ObjC object instantiation in JavaScriptEvaluationResult conversion to JSValueRef
https://bugs.webkit.org/show_bug.cgi?id=291116
rdar://148630517

Reviewed by Ryosuke Niwa.

This not only removes unnecessary work in the conversion process, but it also makes the code
able to be used on platforms that don't have ObjC.

Covered by many tests, especially AsyncFunction API tests and EvaluateJavaScript.ReturnTypes.

* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WTF::DefaultHash<JSC::Strong<JSC::JSCell>>::hash):
(WTF::DefaultHash<JSC::Strong<JSC::JSCell>>::equal):
(WebKit::JavaScriptEvaluationResult::toVariant):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::jsValueToVariant): Deleted.

Canonical link: https://commits.webkit.org/297138@main
